consensus/clique/clique.go

```diff
@@ -0,0 +1,592 @@
+
+	// ErrInvalidVote is returned if a nonce value is something else that the two
+	// allowed constants of 0x00..0 or 0xff..f.
+	ErrInvalidVote = errors.New("vote nonce not 0x00..0 or 0xff..f")
+
+	// ErrInvalidCheckpointVote is returned if a checkpoint/epoch transition block
+	// has a vote nonce set to non zeroes.
+	ErrInvalidCheckpointVote = errors.New("vote nonce in checkpoint block non-zero")
+
+	// ErrMissingVanity is returned if a block's extra-data section is shorter than
+	// 32 bytes, which is required to store the signer vanity.
+	ErrMissingVanity = errors.New("extra-data 32 byte vanity prefix missing")
+
+	// ErrMissingSignature is returned if a block's extra-data section doesn't seem
+	// to contain a 65 byte secp256k1 signature.
+	ErrMissingSignature = errors.New("extra-data 65 byte suffix signature missing")
+
+	// ErrExtraSigners is returned if non-checkpoint block contain signer data in
+	// their extra-data fields.
+	ErrExtraSigners = errors.New("non-checkpoint block contains extra signer list")
+
+	// ErrInvalidCheckpointSigners is returned if a checkpoint block contains an
+	// invalid list of signers (i.e. non divisible by 20 bytes, or not the correct
+	// ones).
+	ErrInvalidCheckpointSigners = errors.New("invalid signer list on checkpoint block")
+
+	// ErrInvalidMixDigest is returned if a block's mix digest is non zero.
+	ErrInvalidMixDigest = errors.New("non-zero mix digest")
+
+	// ErrInvalidUncleHash is returned if a block contains an non-empty uncle list.
+	ErrInvalidUncleHash = errors.New("non empty uncle hash")
+
+	// ErrInvalidDifficulty is returned if the difficulty of a block is not either
+	// of 1 or 2, or if the value does not match the turn of the signer.
+	ErrInvalidDifficulty = errors.New("invalid difficulty")
+
+	// ErrInvalidTimestamp is returned if the timestamp of a block is lower than
+	// the previous block's timestamp + the minimum block period.
+	ErrInvalidTimestamp = errors.New("invalid timestamp")
+
+	// ErrInvalidVotingChain is returned if an authorization list is attempted to
+	// be modified via out-of-range or non-contiguous headers.
+	ErrInvalidVotingChain = errors.New("invalid voting chain")
+
+	// ErrUnauthorized is returned if a header is signed by a non authorized entity.
+	ErrUnauthorized = errors.New("unauthorized")
+)
+
+// SignerFn is a signer callback function to request a hash to be signed by a
+// backing account.
+type SignerFn func(accounts.Account, []byte) ([]byte, error)
+
+// signature is a tiny helper struct to cache recent block headers along with
+// the resolved signers for faster sealing lookups.
+type signature struct {
+	header *types.Header
+	signer common.Address
+}
+
+// ecrecover extracts the Ethereum account address from a signed header.
+func ecrecover(header *types.Header) (common.Address, error) {
+	// Retrieve the signature from the header extra-data
+	if len(header.Extra) < extraSeal {
```

Comment:
The contents of the extra fields are fully validate in `ValidateHeader`. This is just a sanity check so that code below doesn't panic.